File: libs/logger/logger.go
package logger

import (
	"os"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
)

// Logger provides structured logging with trace ID support
type Logger struct {
	logger  zerolog.Logger
	traceID string
}

// GetTraceID returns the current trace ID
func (l *Logger) GetTraceID() string {
	return l.traceID
}

// Config holds basic logger configuration
type Config struct {
	Level  string `json:"level"`  // debug, info, warn, error
	Format string `json:"format"` // json, console
}

// DefaultConfig returns default configuration
func DefaultConfig() Config {
	return Config{
		Level:  "info",
		Format: "json",
	}
}

var globalLogger *Logger

// Init initializes the global logger
func Init(config Config) *Logger {
	level, err := zerolog.ParseLevel(config.Level)
	if err != nil {
		level = zerolog.InfoLevel
	}
	zerolog.SetGlobalLevel(level)

	var logger zerolog.Logger
	if config.Format == "console" {
		logger = log.Output(zerolog.ConsoleWriter{Out: os.Stdout, TimeFormat: time.RFC3339})
	} else {
		logger = log.Logger
	}

	globalLogger = &Logger{
		logger: logger,
	}
	return globalLogger
}

// GetLogger returns the global logger
func GetLogger() *Logger {
	if globalLogger == nil {
		globalLogger = Init(DefaultConfig())
	}
	return globalLogger
}

// WithTraceID creates a new logger with trace ID
func (l *Logger) WithTraceID(traceID string) *Logger {
	return &Logger{
		logger:  l.logger.With().Str("trace_id", traceID).Logger(),
		traceID: traceID,
	}
}

// NewTraceID generates a new trace ID and returns logger with it
func (l *Logger) NewTraceID() *Logger {
	traceID := generateTraceID()
	return l.WithTraceID(traceID)
}

// Debug logs debug message with fixed JSON structure
func (l *Logger) Debug(component, action, message string, data interface{}) {
	l.logger.Debug().
		Str("level", "debug").
		Str("component", component).
		Str("action", action).
		Str("message", message).
		Interface("data", data).
		Time("timestamp", time.Now()).
		Msg("")
}

// Info logs info message with fixed JSON structure
func (l *Logger) Info(component, action, message string, data interface{}) {
	l.logger.Info().
		Str("level", "info").
		Str("component", component).
		Str("action", action).
		Str("message", message).
		Interface("data", data).
		Time("timestamp", time.Now()).
		Msg("")
}

// Warn logs warning message with fixed JSON structure
func (l *Logger) Warn(component, action, message string, data interface{}) {
	l.logger.Warn().
		Str("level", "warn").
		Str("component", component).
		Str("action", action).
		Str("message", message).
		Interface("data", data).
		Time("timestamp", time.Now()).
		Msg("")
}

// Error logs error message with fixed JSON structure
func (l *Logger) Error(component, action, message string, err error, data interface{}) {
	l.logger.Error().
		Str("level", "error").
		Str("component", component).
		Str("action", action).
		Str("message", message).
		Err(err).
		Interface("data", data).
		Time("timestamp", time.Now()).
		Msg("")
}

// Package level functions using global logger

// WithTraceID creates a new logger with trace ID using global logger
func WithTraceID(traceID string) *Logger {
	return GetLogger().WithTraceID(traceID)
}

// NewTraceID generates a new trace ID using global logger
func NewTraceID() *Logger {
	return GetLogger().NewTraceID()
}

// Debug logs debug message using global logger
func Debug(component, action, message string, data interface{}) {
	GetLogger().Debug(component, action, message, data)
}

// Info logs info message using global logger
func Info(component, action, message string, data interface{}) {
	GetLogger().Info(component, action, message, data)
}

// Warn logs warning message using global logger
func Warn(component, action, message string, data interface{}) {
	GetLogger().Warn(component, action, message, data)
}

// Error logs error message using global logger
func Error(component, action, message string, err error, data interface{}) {
	GetLogger().Error(component, action, message, err, data)
}

// generateTraceID generates a simple trace ID
func generateTraceID() string {
	return time.Now().Format("20060102150405") + "-" + randomString(8)
}

// randomString generates a random string of given length
func randomString(length int) string {
	const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[time.Now().UnixNano()%int64(len(charset))]
	}
	return string(b)
}

File: libs/logger/logger.ts
export interface LogData {
  [key: string]: any;
}

export interface LogConfig {
  level: 'debug' | 'info' | 'warn' | 'error';
  format: 'json' | 'console';
}

interface LogEntry {
  level: string;
  trace_id?: string;
  component: string;
  action: string;
  message: string;
  data?: LogData;
  error?: string;
  timestamp: string;
}

export class Logger {
  private traceId?: string;
  private config: LogConfig;

  constructor(config: LogConfig, traceId?: string) {
    this.config = config;
    this.traceId = traceId;
  }

  /**
   * Creates a new logger with trace ID
   */
  withTraceId(traceId: string): Logger {
    return new Logger(this.config, traceId);
  }

  /**
   * Generates a new trace ID and returns logger with it
   */
  newTraceId(): Logger {
    const traceId = this.generateTraceId();
    return new Logger(this.config, traceId);
  }

  /**
   * Logs debug message with fixed JSON structure
   */
  debug(component: string, action: string, message: string, data?: LogData): void {
    if (this.shouldLog('debug')) {
      this.writeLog('debug', component, action, message, undefined, data);
    }
  }

  /**
   * Logs info message with fixed JSON structure
   */
  info(component: string, action: string, message: string, data?: LogData): void {
    if (this.shouldLog('info')) {
      this.writeLog('info', component, action, message, undefined, data);
    }
  }

  /**
   * Logs warning message with fixed JSON structure
   */
  warn(component: string, action: string, message: string, data?: LogData): void {
    if (this.shouldLog('warn')) {
      this.writeLog('warn', component, action, message, undefined, data);
    }
  }

  /**
   * Logs error message with fixed JSON structure
   */
  error(component: string, action: string, message: string, error?: Error, data?: LogData): void {
    if (this.shouldLog('error')) {
      this.writeLog('error', component, action, message, error, data);
    }
  }

  private shouldLog(level: string): boolean {
    const levels = ['debug', 'info', 'warn', 'error'];
    const configLevelIndex = levels.indexOf(this.config.level);
    const logLevelIndex = levels.indexOf(level);
    return logLevelIndex >= configLevelIndex;
  }

  private writeLog(
    level: string,
    component: string,
    action: string,
    message: string,
    error?: Error,
    data?: LogData
  ): void {
    const entry: LogEntry = {
      level,
      component,
      action,
      message,
      timestamp: new Date().toISOString(),
    };

    if (this.traceId) {
      entry.trace_id = this.traceId;
    }

    if (error) {
      entry.error = error.message;
    }

    if (data) {
      entry.data = data;
    }

    if (this.config.format === 'console') {
      this.writeConsoleLog(entry);
    } else {
      this.writeJsonLog(entry);
    }
  }

  private writeConsoleLog(entry: LogEntry): void {
    const timestamp = new Date(entry.timestamp).toLocaleTimeString();
    const traceInfo = entry.trace_id ? `[${entry.trace_id}] ` : '';
    const levelColor = this.getLevelColor(entry.level);
    
    console.log(
      `${timestamp} ${levelColor}${entry.level.toUpperCase()}${this.resetColor()} ${traceInfo}${entry.component}:${entry.action} - ${entry.message}`,
      entry.data ? entry.data : '',
      entry.error ? `Error: ${entry.error}` : ''
    );
  }

  private writeJsonLog(entry: LogEntry): void {
    console.log(JSON.stringify(entry));
  }

  private getLevelColor(level: string): string {
    if (typeof window !== 'undefined') return ''; // Browser environment
    
    const colors = {
      debug: '\x1b[36m', // Cyan
      info: '\x1b[32m',  // Green
      warn: '\x1b[33m',  // Yellow
      error: '\x1b[31m', // Red
    };
    return colors[level as keyof typeof colors] || '';
  }

  private resetColor(): string {
    if (typeof window !== 'undefined') return ''; // Browser environment
    return '\x1b[0m';
  }

  private generateTraceId(): string {
    const timestamp = new Date().toISOString().replace(/[-:.]/g, '').slice(0, 14);
    const random = this.randomString(8);
    return `${timestamp}-${random}`;
  }

  private randomString(length: number): string {
    const charset = 'abcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
      result += charset.charAt(Math.floor(Math.random() * charset.length));
    }
    return result;
  }
}

// Default configuration
export const defaultConfig: LogConfig = {
  level: 'info',
  format: 'json',
};

// Global logger instance
let globalLogger: Logger;

/**
 * Initializes the global logger with configuration
 */
export function init(config: LogConfig = defaultConfig): Logger {
  globalLogger = new Logger(config);
  return globalLogger;
}

/**
 * Gets the global logger instance
 */
export function getLogger(): Logger {
  if (!globalLogger) {
    globalLogger = new Logger(defaultConfig);
  }
  return globalLogger;
}

// Package level functions using global logger

/**
 * Creates a new logger with trace ID using global logger
 */
export function withTraceId(traceId: string): Logger {
  return getLogger().withTraceId(traceId);
}

/**
 * Generates a new trace ID using global logger
 */
export function newTraceId(): Logger {
  return getLogger().newTraceId();
}

/**
 * Logs debug message using global logger
 */
export function debug(component: string, action: string, message: string, data?: LogData): void {
  getLogger().debug(component, action, message, data);
}

/**
 * Logs info message using global logger
 */
export function info(component: string, action: string, message: string, data?: LogData): void {
  getLogger().info(component, action, message, data);
}

/**
 * Logs warning message using global logger
 */
export function warn(component: string, action: string, message: string, data?: LogData): void {
  getLogger().warn(component, action, message, data);
}

/**
 * Logs error message using global logger
 */
export function error(component: string, action: string, message: string, error?: Error, data?: LogData): void {
  getLogger().error(component, action, message, error, data);
}
